### PR TITLE
Setoid rewrite: do so that the internal do_subrelation local definition do not escape its scope (subsumed by #14850)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** tentative fix

For some reason, suspectingly to simulate a form of `Prop`/`Type` polymorphism, `setoid rewrite` internally works in an environment extended with a local definition named `do_subrelation` which binds to either `Morphisms.do_subrelation` of `CMorphisms.do_subrelation`. However, this local definition was able to escape its scope, leading in the original env to trivial unification problems of the form `do_subrelation == Morphisms.do_subrelation` which could not be solved any longer.

This PR ensures that when coming back to the original environment which does not have `do_subrelation`, the remaining occurrences are unfolded.

I'm not sure this is the best imaginable fix but it seems like a reasonable patch to the `do_subrelation` trick.

For a better parenthesizing of when `do_subrelation` is added then unfolded, I moved the place where it was added, leading a priori to more evars depending by default on this local definition. I have a small interrogation about whether this can modify the behavior of unification at some places internally to `rewrite` but it seems to work.

This was revealed by #14673 where an equation `?x@{do_subrelation} == ?x{Morphisms.do_subrelation}` was wrongly unresolved (due to the `rewrite` bug) but nevertheless restricted (thanks to #14673) w/o the unification heuristics being able to see that some progress had been made (yet another limitation to fix at some time).

### CODE REVIEW

File: plugins/ltac/rewrite.ml [hunk 1]
```diff
@@ -1476,7 +1481,8 @@ let cl_rewrite_clause_aux ?(abs=None) strat env avoid sigma concl is_hyp : resul
                 Termops.pr_evar_info env evars (Evd.find evars ev)))
         cstrs
     in
-    let newt = res.rew_to in
+    let subst = [do_subrelation_id, fst (do_subrelation_decl env res.rew_evars prop)] in
```
Comment (ppedrot) [inline plugins/ltac/rewrite.ml:1484]: Are we sure that we're going to unify the universe instances of the freshly generated constant in universe-poly mode? Before we only had one but now we get one for the let-bound variable and one for the substituted term, which might lead to unification failures.

Comment (herbelin) [reply plugins/ltac/rewrite.ml:1484]: Good question.

The universes for `apply_subrelation` and `do_subrelation` were thrown before, but this is defined in `Prop` (`CoInductive apply_subrelation : Prop := do_subrelation.` in both `Morphisms.v` and `CMorphisms.v`), so universes seem to not matter. If confirmed, this is a worth a remark though.

**Comment by SkySkimmer** [comment]
This PR/commit title is not very grammatical, I cannot parse it.

**Comment by herbelin** [comment]
> This PR/commit title is not very grammatical, I cannot parse it.

I intended something like: "Faire en sorte que la définition locale do_subrelation n'échappe pas la portée du lieur qui l'introduit".

It would probably be more intuitive to simply not add this local definition in the first place but I don't understand well its purpose and what could be an alternative implementation.

**Comment by SkySkimmer** [comment]
"Ensure that the internal do_subrelation local definition does not escape its scope" should work then

**Comment by herbelin** [comment]
> "Ensure that the internal do_subrelation local definition does not escape its scope" should work then

To be honest, "ensure" is maybe a bit too strong. I don't have a general overview of this code enough to claim that all occurrences are eventually substituted. For instance, some remain maybe in conversion problems or in other data used by the the tactic. Getting rid of the do_subrelation "hack" in the first place would probably be more satisfactory.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Bench just in case: https://gitlab.com/coq/coq/-/jobs/1476050474

**Comment by ppedrot** [comment]
I don't know whether the verdi slowdown is real, we should probably relaunch a bench on this dev only...

**Comment by ppedrot** [comment]
Rerunning: https://gitlab.com/coq/coq/-/jobs/1504759418

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-parsers │  579.97   584.69  -0.81 │  1624270898498   1631514012258  -0.44 │  2485513467263   2506518516547  -0.84 │ 3084952  3084872   0.00 │   0    0    nan │
│                 coq-bignums │   55.82    56.25  -0.76 │   154355104796    155069723905  -0.46 │   207098900858    207091264588   0.00 │  463512   463496   0.00 │   0    0    nan │
│       coq-mathcomp-fingroup │   48.21    48.51  -0.62 │   133884864728    134021848094  -0.10 │   186526161619    186526274880  -0.00 │  480448   480452  -0.00 │   0    0    nan │
│              coq-coquelicot │   68.58    68.95  -0.54 │   187004083074    187109117342  -0.06 │   240534231685    239581882295   0.40 │  557248   557528  -0.05 │   0    0    nan │
│                   coq-color │  481.30   483.58  -0.47 │  1345538917992   1346521445351  -0.07 │  1697899483092   1696516679186   0.08 │ 1154960  1155020  -0.01 │   0    0    nan │
│       coq-engine-bench-lite │  290.67   291.68  -0.35 │   818519349514    821953770836  -0.42 │  1494114592582   1497177837500  -0.20 │ 1411800  1411876  -0.01 │   0    0    nan │
│       coq-mathcomp-solvable │  169.38   169.92  -0.32 │   471596710082    472586835835  -0.21 │   665801151146    665875177466  -0.01 │  693600   691768   0.26 │   0    0    nan │
│      coq-mathcomp-character │  149.87   150.33  -0.31 │   417000138889    417848458867  -0.20 │   583495785693    583526594946  -0.01 │  733220   732864   0.05 │   0    0    nan │
│                    coq-corn │ 1457.72  1461.04  -0.23 │  4062680426196   4067616737787  -0.12 │  5864310855097   5863113392889   0.02 │  790632   790624   0.00 │   0    0    nan │
│      coq-mathcomp-ssreflect │   53.75    53.86  -0.20 │   147258805886    147523975893  -0.18 │   191069098527    191038304647   0.02 │  527396   527312   0.02 │   0    0    nan │
│             coq-fiat-crypto │ 4842.38  4850.06  -0.16 │ 13445636782560  13471341244928  -0.19 │ 23756624657719  23754253009007   0.01 │ 2228640  2225908   0.12 │   0    0    nan │
│                  coq-stdlib │  412.28   412.84  -0.14 │  1152076669166   1155514180133  -0.30 │  1467375792656   1467171327347   0.01 │  602616   602612   0.00 │   0    0    nan │
│                   coq-flocq │  146.10   146.29  -0.13 │   404809791558    405880328689  -0.26 │   503841108311    503761507320   0.02 │  810260   810264  -0.00 │   0    0    nan │
│               coq-perennial │ 7204.74  7210.12  -0.07 │ 20065277907163  20084789456241  -0.10 │ 30707442427190  30693774383848   0.04 │ 2979616  3008476  -0.96 │   0    0    nan │
│  coq-performance-tests-lite │ 1546.63  1547.30  -0.04 │  4318207188580   4315447990394   0.06 │  7121681320158   7116112595295   0.08 │ 2085088  2085120  -0.00 │   0    0    nan │
│                coq-rewriter │  686.37   686.65  -0.04 │  1904178232873   1902910598264   0.07 │  2932114183367   2931456524110   0.02 │ 1074684  1118496  -3.92 │   0    0    nan │
│               coq-fourcolor │ 2546.50  2546.76  -0.01 │  7098401046710   7098169592153   0.00 │ 12234040766324  12227861508656   0.05 │  732540   732672  -0.02 │   0    0    nan │
│                  coq-geocoq │ 1321.84  1321.94  -0.01 │  3684269644468   3683884646901   0.01 │  5419524540724   5417939087434   0.03 │  982628   977288   0.55 │   0    0    nan │
│                coq-coqprime │  326.00   325.88   0.04 │   905931674457    905639406660   0.03 │  1595985541503   1595100250813   0.06 │  918148   918448  -0.03 │   0    0    nan │
│      coq-mathcomp-odd-order │ 1047.17  1046.76   0.04 │  2914248611743   2915535105338  -0.04 │  4623047234866   4622862001505   0.00 │  908152   942528  -3.65 │   0    0    nan │
│            coq-math-classes │  172.52   172.42   0.06 │   480885345051    481270401012  -0.08 │   641214703221    641061106713   0.02 │  481960   482008  -0.01 │   0    0    nan │
│                 coq-unimath │ 4665.88  4661.12   0.10 │ 12980528676664  12973928219442   0.05 │ 24872580866967  24877392847270  -0.02 │ 3758348  3725688   0.88 │   0    0    nan │
│                    coq-hott │  296.99   296.62   0.12 │   815528898499    815569608831  -0.00 │  1267488923230   1267563622100  -0.01 │  570912   570876   0.01 │   0    0    nan │
│                    coq-core │  107.40   107.21   0.18 │   298290213601    297921955585   0.12 │   461658865727    461509292160   0.03 │  298836   298860  -0.01 │   0    0    nan │
│ coq-rewriter-perf-SuperFast │ 1637.50  1634.47   0.19 │  4555233870071   4550264733659   0.11 │  7151971011092   7150622113253   0.02 │ 1074388  1074464  -0.01 │   0    0    nan │
│                coq-bedrock2 │  302.45   301.76   0.23 │   837909197189    837408018301   0.06 │  1493921148039   1493558674564   0.02 │  780700   800024  -2.42 │   0    0    nan │
│          coq-mathcomp-field │  222.32   221.81   0.23 │   618062142592    617240744337   0.13 │   938548748017    938561034523  -0.00 │  665172   665156   0.00 │   0    0    nan │
│        coq-mathcomp-algebra │  141.46   140.90   0.40 │   392674876191    392234796255   0.11 │   510007564392    510076700075  -0.01 │  557252   557000   0.05 │   0    0    nan │
│              coq-verdi-raft │ 1100.56  1096.04   0.41 │  3064069942176   3053382186426   0.35 │  4305725859527   4298905461461   0.16 │  920928   920416   0.06 │   0    0    nan │
│                 coq-coqutil │   60.08    59.29   1.33 │   165606162010    164642633525   0.59 │   218732591575    218183956695   0.25 │  599968   599776   0.03 │   0    0    nan │
│                   coq-verdi │   98.54    95.50   3.18 │   270839264352    263915839717   2.62 │   379343608556    373622713566   1.53 │  542072   542748  -0.12 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```
So it does look like verdi is slowed down by this PR.

**Comment by herbelin** [comment]
> So it does look like verdi is slowed down by this PR.

I don't see what else than the `replace_vars` can be in cause. But this `replace_vars` is necessary to prevent the scope escaping.

Anyway, #14673 does not need it anymore. So, maybe can we wait for @mattam82's feedback. Maybe would there be after all a more principled way to deal with the `Morphisms.do_subrelation`/`CMorphisms.do_subrelation` distinction.

**Comment by ppedrot** [comment]
I had another look at this PR and I'm now wondering why it's done this way.

Even without considering the fix, `do_subrelation` could be the same for Prop / Type rewriting. It's just a marker that tracks the resolution phase of the rewriting algorithm. There are a few more like that, e.g. `normalization_done`.

Second, the only place where the variable is reachable is the `morph` evar found in this snippet.
```
    let env' =
      let dosub, appsub =
        if b then PropGlobal.do_subrelation, PropGlobal.apply_subrelation
        else TypeGlobal.do_subrelation, TypeGlobal.apply_subrelation
      in
        EConstr.push_named
          (LocalDef (make_annot (Id.of_string "do_subrelation") Sorts.Relevant,
                     snd (app_poly_sort b env evars dosub [||]),
                     snd (app_poly_nocheck env evars appsub [||])))
          env
    in
    let evars, morph = new_cstr_evar evars env' app in
      evars, morph, morph, sigargs, appm, morphobjs, morphobjs'
```
Why don't we simply expand it in the evar instance, replacing it with the global definition?

**Comment by ppedrot** [comment]
@herbelin do you have a test-case for the bug this PR is fixing?

**Comment by herbelin** [comment]
> @herbelin do you have a test-case for the bug this PR is fixing?

No, I observed the scope escaping while experimenting with #14673 but I finally made that #14673 does not depend on this PR.

**Comment by ppedrot** [comment]
Here is a hackish test.
```
From Coq Require Import Setoid Morphisms.

Axiom T : Type.
Axiom Teq : relation T.
#[local] Declare Instance Teq_Equivalence : Equivalence Teq.

Axiom P : T -> Prop.
Axiom P_Proper : Proper (Teq ==> Basics.flip Basics.impl) P.

Ltac manual_P_proper :=
match goal with
| [ H : apply_subrelation |- Proper (Teq ==> Basics.flip Basics.impl) P ] =>
  case H; apply P_Proper
end.

#[local] Hint Extern 1 => manual_P_proper : typeclass_instances.

Lemma foo : forall x y : T, Teq x y -> P y -> P x.
Proof.
intros m n e h.
rewrite e.
exact h.
Qed.
```

**Comment by ppedrot** [comment]
@herbelin see https://github.com/coq/coq/pull/14850 for a more local fix, which is likely to be more efficient as well.

**Pull request closed without merging**